### PR TITLE
deal with bad chosen representations of k-regular sequences

### LINKED ISSUE #21343: deal with bad chosen representations of k-regular sequences
<div id="comment:0"></div>

Deal with the troubles coming from bad chosen representations of k-regular sequences, where `mu[0]*right != right`. (This makes troubles as leading zeros of digit expansions should not influence the result.)

Depends on #21204

Component: **combinatorics**

Author: **Daniel Krenn**

Branch/Commit: **[u/dkrenn/k-regular-warning](https://github.com/sagemath/sagetrac-mirror/tree/u/dkrenn/k-regular-warning) @ [`4af416c`](https://github.com/sagemath/sagetrac-mirror/commit/4af416c4e40895ebd54f477464839d33804730f8)**

_Issue created by migration from https://trac.sagemath.org/ticket/21343_

### PR DESCRIPTION
**Pull request by dkrenn**
### :books: Description

It deals with the situation when mu[0]*right != right in k-regular sequences.

Fixes #21343. This PR is created from the branch/code that has been on the corresponding trac ticket + merging in a current SageMath version.

See also meta issue #21202.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### :hourglass: Dependencies

- #21204: guessing k-regular sequences from the first few values

No formal dependencies, but there are trivial merge conflicts---in both branches, methods are inserted at the same position in the code---between:
- #21343: deal with mu[0]*right != right in k-regular sequences
- #35894: Regular sequences: implement convolution / ring structure

### CODE REVIEW

File: src/sage/combinat/k_regular_sequence.py [hunk 1]
```diff
@@ -55,7 +59,8 @@
 There is a `2`-recursive sequence describing the numbers above as well::
 
     sage: U = Seq2((Matrix([[3, 2], [0, 1]]), Matrix([[2, 0], [1, 3]])),
-    ....:          left=vector([0, 1]), right=vector([1, 0])).transposed()
+    ....:          left=vector([0, 1]), right=vector([1, 0]),
+    ....:          allow_degenerated_sequence=True).transposed()
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]: I would prefer a non-degenerated example as the second overall example in this file, and perhaps we could skip `transposed` also when rewriting this.

File: src/sage/combinat/k_regular_sequence.py [hunk 2]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+
+        OUTPUT:
+
+        A :class:`kRegularSequence`
+
+        EXAMPLES::
+
+            sage: Seq2 = kRegularSequenceSpace(2, ZZ)
+
+        The following linear representation of `S` is chosen bad (is
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]:
```suggestion
        The following linear representation of `S` is chosen badly (is
```

File: src/sage/combinat/k_regular_sequence.py [hunk 3]
```diff
@@ -413,6 +730,27 @@ def subsequence(self, a, b):
             Traceback (most recent call last):
             ...
             ValueError: a=-1 is not nonnegative.
+
+        The following linear representation of `S` is chosen bad (is
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]:
```suggestion
        The following linear representation of `S` is chosen badly (is
```

File: src/sage/combinat/k_regular_sequence.py [hunk 4]
```diff
@@ -719,6 +1057,55 @@ def partial_sums(self, include_n=False):
             sage: C.partial_sums(include_n=True)
             2-regular sequence 0, 1, 3, 6, 10, 15, 21, 28, 36, 45, ...
 
+        The following linear representation of `S` is chosen bad (is
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]:
```suggestion
        The following linear representation of `S` is chosen badly (is
```

File: src/sage/combinat/k_regular_sequence.py [hunk 5]
```diff
@@ -918,6 +1305,510 @@ def _n_to_index_(self, n):
...
+            sage: S.linear_representation()
+            ((1, 0),
+             Finite family {0: [0 0]
+                               [0 1],
+                            1: [0 1]
+                               [0 1]},
+             (0, 1))
+            sage: logging.getLogger().setLevel(logging.WARN)
+
+        The following linear representation of `S` is chosen bad (is
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]:
```suggestion
        The following linear representation of `S` is chosen badly (is
```

File: src/sage/combinat/k_regular_sequence.py [hunk 6]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+            sage: S.is_degenerated()
+            True
+
+        ::
+
+            sage: C = Seq2((Matrix([[2, 0], [2, 1]]), Matrix([[0, 1], [-2, 3]])),
+            ....:          vector([1, 0]), vector([0, 1]))
+            sage: C.is_degenerated()
+            False
+        """
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:381]: I suggest to add a `SEEALSO` block linking to `regenerated`.

File: src/sage/combinat/k_regular_sequence.py [hunk 7]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+                "You can use 'allow_degenerated_sequence=True' followed by "
+                "a call of method .regenerated() "
+                "for correcting this.")
+
+    @cached_method
+    @minimize_result
+    def regenerated(self):
+        r"""
+        Return a `k`-regular sequence that satisfies
+        `\mu[0] \mathit{right} = \mathit{right}`.
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]:
```suggestion
        `\mu[0] \mathit{right} = \mathit{right}` with the same values as
        this sequence.
```

File: src/sage/combinat/k_regular_sequence.py [hunk 8]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+        z = next(itA)
+        mu = {z: Matrix.block([[I, Z], [Z, self.mu[z]]])}
+        mu.update((r, Matrix.block([[Z, Z], [self.mu[r], self.mu[r]]]))
+                  for r in itA)
+
+        return P.element_class(
+            P, mu,
+            vector(2*tuple(self.left)),
+            vector(tuple(self.right) + dim*(0,)))
+
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:505]: This version here constructs a `2*dim`-dimensional linear representation. The corresponding right vector valued sequence $\tilde v(n)$ is $(v(n)[n=0], v(n)[n\neq 0])^\top$ where $v(n)$ is the right vector valued sequence of the original linear representation.

Theorem B of [Heuberger, C., Krenn, D. & Lipnik, G.F. Asymptotic Analysis of q-Recursive Sequences. Algorithmica 84, 2480–2532 (2022)](https://doi.org/10.1007/s00453-022-00950-y) constructs a $dim+1$-dimensional linear representation at no significantly higher cost.

Therefore, I suggest to 
- replace the implementation here by the one of the above mentioned paper;
- write a brief Algorithm section for the docstring which refers to that paper.

File: src/sage/combinat/k_regular_sequence.py [hunk 9]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+        EXAMPLES::
+
+            sage: Seq2 = kRegularSequenceSpace(2, ZZ)
+            sage: U = Seq2((Matrix([[3, 2], [0, 1]]), Matrix([[2, 0], [1, 3]])),
+            ....:          left=vector([0, 1]), right=vector([1, 0]),
+            ....:          allow_degenerated_sequence=True)
+            sage: U.is_degenerated()
+            True
+            sage: Ut = U.transposed()
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:533]: I suggest to show the linear representation of `Ut` here (for readers who are not fully acquainted with `.transposed()` as a recognizable series.

File: src/sage/combinat/k_regular_sequence.py [hunk 10]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+            Using such a sequence might lead to wrong results.
+            You can use 'allow_degenerated_sequence=True' followed
+            by a call of method .regenerated() for correcting this.
+            sage: Utt = Ut.transposed(allow_degenerated_sequence=True)
+            sage: Utt.is_degenerated()
+            True
+
+        .. SEEALSO::
+
+            :meth:`RecognizableSeries.tranposed <sage.combinat.recognizable_series.RecognizableSeries.tranposed>`
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]: The link is broken here.

File: src/sage/combinat/k_regular_sequence.py [hunk 11]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+        Return a recognizable series equivalent to this series, but
+        with a right minimized linear representation.
+
+        OUTPUT:
+
+        A :class:`kRegularSequence`
+
+        .. SEEALSO::
+
+            :meth:`RecognizableSeries._minimized_right_ <sage.combinat.recognizable_series.RecognizableSeries._minimized_right>`
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]: Link seems to be broken.

File: src/sage/combinat/k_regular_sequence.py [hunk 12]
```diff
@@ -240,6 +336,227 @@ def __iter__(self):
...
+            :meth:`RecognizableSeries.tranposed <sage.combinat.recognizable_series.RecognizableSeries.tranposed>`
+        """
+        element = super().transposed()
+        if not allow_degenerated_sequence:
+            element._error_if_degenerated_()
+        return element
+
+    def _minimized_right_(self):
+        r"""
+        Return a recognizable series equivalent to this series, but
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]:
```suggestion
        Return a regular sequence equivalent to this series, but
```

**Review by cheuberg — CHANGES_REQUESTED** [review]
Here are my remarks on this PR.

**Review by cheuberg — CHANGES_REQUESTED** [review]
(basically just a marker that I reviewed the last changes)

File: src/sage/combinat/k_regular_sequence.py (hunk 10 above)
Comment (dkrenn) [reply src/sage/combinat/k_regular_sequence.py]: 48efc431f0c

File: src/sage/combinat/k_regular_sequence.py (hunk 11 above)
Comment (dkrenn) [reply src/sage/combinat/k_regular_sequence.py]: 48efc431f0c

File: src/sage/combinat/k_regular_sequence.py (hunk 9 above)
Comment (dkrenn) [reply src/sage/combinat/k_regular_sequence.py:533]: dc4e55ca758

File: src/sage/combinat/k_regular_sequence.py (hunk 1 above)
Comment (dkrenn) [reply src/sage/combinat/k_regular_sequence.py]: 072a82e94ee

File: src/sage/combinat/k_regular_sequence.py (hunk 8 above)
Comment (dkrenn) [reply src/sage/combinat/k_regular_sequence.py:505]: Indeed, this should be changed. I'll prepare a patch...

File: src/sage/combinat/k_regular_sequence.py (hunk 1 above)
Comment (cheuberg) [reply src/sage/combinat/k_regular_sequence.py]: Thank you, please resolve.

File: src/sage/combinat/k_regular_sequence.py (hunk 9 above)
Comment (cheuberg) [reply src/sage/combinat/k_regular_sequence.py:533]: Thank you, please resolve.

File: src/sage/combinat/k_regular_sequence.py (hunk 10 above)
Comment (cheuberg) [reply src/sage/combinat/k_regular_sequence.py]: Thank you, please resolve.

File: src/sage/combinat/k_regular_sequence.py (hunk 11 above)
Comment (cheuberg) [reply src/sage/combinat/k_regular_sequence.py]: Thank you, please resolve.

**Review by cheuberg — CHANGES_REQUESTED** [review]
Changes LGTM, open discussion(s)

File: src/sage/combinat/k_regular_sequence.py (hunk 8 above)
Comment (dkrenn) [reply src/sage/combinat/k_regular_sequence.py:505]: d665cd6d31e

**Comment by dkrenn** [comment]
> Changes LGTM, open discussion(s)

I've dealt with all open discussions; please review.

**Review by cheuberg — APPROVED** [review]
Thank you, LGTM. The reference [HKL2021] did appear, I prepared a PR to be pulled into this branch here: https://github.com/dkrenn/sage/pull/2

**Review by cheuberg — APPROVED** [review]
From my point of view, this PR is ready to be merged. @dkrenn, can you please set appropriate labels?

**Comment by dkrenn** [comment]
Failing checks seem to be unrelated to the changes of this PR: incremental build fails with some `OSError` at `matlibplot`, Lint/RST in `schemes.projective`, codecov only brings up false-positives.

**Pull request merged**